File: Evaluate Reverse Polish Notation (RPN) - Leetcode 150/Evaluate Reverse Polish Notation (RPN) - Leetcode 150.java
import java.util.Stack;

public class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();

        for (String token : tokens) {
            if (token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/")) {
                int b = stack.pop();
                int a = stack.pop();
                
                switch (token) {
                    case "+":
                        stack.push(a + b);
                        break;
                    case "-":
                        stack.push(a - b);
                        break;
                    case "*":
                        stack.push(a * b);
                        break;
                    case "/":
                        double division = (double) a / b;
                        stack.push(division < 0 ? (int) Math.ceil(division) : (int) Math.floor(division));
                        break;
                }
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        
        return stack.pop();
    }
}

// Time Complexity: O(n)
// Space Complexity: O(n)

File: Evaluate Reverse Polish Notation (RPN) - Leetcode 150/Evaluate Reverse Polish Notation (RPN) - Leetcode 150.js
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function(tokens) {
    let stack = [];

    for (let token of tokens) {
        if (token === '+' || token === '-' || token === '*' || token === '/') {
            let b = stack.pop();
            let a = stack.pop();
            
            switch (token) {
                case '+':
                    stack.push(a + b);
                    break;
                case '-':
                    stack.push(a - b);
                    break;
                case '*':
                    stack.push(a * b);
                    break;
                case '/':
                    let division = a / b;
                    stack.push(division < 0 ? Math.ceil(division) : Math.floor(division));
                    break;
            }
        } else {
            stack.push(parseInt(token));
        }
    }
    
    return stack[0];
};

// Time Complexity: O(n)
// Space Complexity: O(n)
